packages/gateway/src/Shard.ts

```diff
@@ -339,17 +427,74 @@ export class DiscordenoShard {
 
   /** Handle an incoming gateway message. */
   async handleMessage(message: MessageEvent): Promise<void> {
-    let preProcessMessage = message.data
+    // The ws npm package will use a Buffer, while the global built-in will use ArrayBuffer
+    const isCompressed = message.data instanceof ArrayBuffer || message.data instanceof Buffer
+
+    const data = isCompressed ? await this.decompressPacket(message.data) : (JSON.parse(message.data) as DiscordGatewayPayload)
+
+    // Check if the decompression was not successful
+    if (!data) return
 
-    // If message compression is enabled, Discord might send zlib compressed payloads.
-    if (this.gatewayConfig.compress && preProcessMessage instanceof Blob) {
-      preProcessMessage = inflateSync(await preProcessMessage.arrayBuffer()).toString()
+    await this.handleDiscordPacket(data)
+  }
+
+  /**
+   * Decompress a zlib/zstd compressed packet
+   *
+   * @private
+   */
+  async decompressPacket(data: ArrayBuffer | Buffer): Promise<DiscordGatewayPayload | null> {
+    // A buffer is a Uint8Array under the hood. An ArrayBuffer is generic, so we need to create the Uint8Array that uses the whole ArrayBuffer
+    const compressedData: Uint8Array = data instanceof Buffer ? data : new Uint8Array(data)
+
+    if (this.gatewayConfig.transportCompression === TransportCompression.zlib) {
```

Comment:
(preference/stylistic, feel free to ignore) maybe switch case? idk I just think it make more sense in this case

Reply:
In this case it does not work, it would be a `switch` (for transport compression) + an `if` later (for payload comperssion), might as well keep it an `if`